src/mozilla_sec_eia/models/sec10k/ex_21/inference.py

```diff
@@ -180,88 +182,140 @@ def get_flattened_mode_predictions(token_boxes_tensor, predictions_tensor):
-        model: A fine-tuned LayoutLM model.
-        processor: The tokenizer and encoder for model inputs.
-        extraction_metadata: A dataframe to track extraction success metrics. Should
-            have columns 'filename' and 'success'.
-        dataset_ind: A list of index numbers of dataset records to be used for inference
-            Default is None, in which the entire dataset created from the PDF directory
-            is used.
-        labeled_json_dir: Path to the directory with labeled JSONs from Label Studio. Cannot
-            be None if has_labels is True.
-        has_labels: Boolean, true if the data has associated labels that can be used in
-            visualizing and validating results.
-        device: String or int, specify what computation device to use for inference
-            i.e. "mps", "cpu", "cuda"
-
-    Returns:
-        logits: A list of logits. The list is the length of the number of documents in the
-            dataset (number of PDFs in pdfs_dir). Each logit object in the list is of
-            shape (batch_size, seq_len, num_labels). Seq_len is
-            the same as token length (512 in this case).
-        predictions: A list of predictions. The list is the length of the number of documents
-            in the dataset (number of PDFs in pdfs_dir).
-            From the logits, we take the highest score for each token, using argmax.
-            This serves as the predicted label for each token. It is shape (seq_len) or token
-            length.
-        output_dfs: The extracted Ex. 21 tables. This is one big dataframe with an ID column
-            that is the filename of the extracted Ex. 21. Dataframe contains columns id,
-            subsidiary, loc, own_per.
-    """
-    dataset = create_inference_dataset(
-        pdfs_dir=pdfs_dir, labeled_json_dir=labeled_json_dir, has_labels=has_labels
-    )
-    if dataset_ind:
-        dataset = dataset.select(dataset_ind)
-
-    # TODO: figure out device argument
-    pipe = pipeline(
-        "token-classification",
-        model=model,
-        tokenizer=processor,
-        pipeline_class=LayoutLMInferencePipeline,
-        device=device,
-    )
+class Exhibit21Extractor(ConfigurableResource):
+    """Implement `Sec10kExtractor` interface for exhibit 21 data."""
+
+    cloud_interface: GCSArchive
+    layoutlm: LayoutlmResource
+    name: str = "exhibit21_extractor"
+    device: str = "cpu"
+    has_labels: bool = False
+    dataset_ind: list | None = None
+
+    @contextmanager
+    def setup_for_execution(self, context):
+        """Set env variable to improve GPU memory access."""
+        os.environ["PYTORCH_CUDA_ALLOC_CONF"] = "expandable_segments:True"
+
+    def extract_filings(
+        self, filing_metadata: pd.DataFrame
+    ) -> tuple[pd.DataFrame, pd.DataFrame]:
+        """Predict entities with a fine-tuned model and extract Ex. 21 tables.
+
+        This function starts by creating a HuggingFace dataset from PDFs in `pdfs_dir`
```

Comment:
context: in my mind, this is what's happening - let me know what's wrong

* we generate a bunch of PDFs, then annotate them and munge that into a huggingface `Dataset` - afaik this is a unified interface that you can pass to any number of huggingface models?
* huggingface has a pretrained layoutlm model that we can just download
* we push the pdfs through the layoutlm model which gets us a dataframe with `id, subsidiary, loc, own_per` columns
* do some cleaning, tada!